Update language-and-grammar.md (#36)
AgoraDoc/agora-style-guide: _content/language-and-grammar.md

   For more information, contace Agora Inc.
```

### Active voice

**Quick reference:**
- Favor the active voice where possible.
_________

Generally speaking, with the active voice the subject of a sentence performs the action, whereas with the passive voice the subject is acted upon by something else, .

The active voice sounds immediate and clear. The passive voice usually sounds more wordy and can obscure the subject, making it difficult for readers to fully understand your meaning. 

| Recommended                       | Not recommended                         |
| :-------------------------------- | :-------------------------------------- |
| This method gets the user ID. | This method is used to get the user ID. |
| The callback provides troubleshooting information | Troubleshooting information is provided by the callback. |

This is not a hard and fast rule. The passive voice can be used to emphasize the object of the action or the action itself. Also, it may not always be possible to phrase your sentence with an active voice, such as in status code explanations. For example, "This method is not implemented."

### Anthropomorphism

**Quick reference:**

> - Do not attribute human qualities to software or hardware.

Author: Probably

| :-------------------------------------------------------- | :----------------------------------------------------------- |
| A Delimiter object specifies where to split a string. | A Delimiter object tells the splitter where a string should be broken. |
| The PC detects a new device.                          | The PC sees a new device.                                    |

### Articles (a, an, the)

**Quick reference:**
- Singular countable nouns must have an article or similar determiner.
- Plural countable nouns can have a definite article or none.
- Do not use an article with proper nouns or uncountable (mass) nouns.
- Use a definite article plus "method" or "callback" to specify its function unless it is obvious from context.
- Do not normally use a leading article with "mode." Example: "Enter silent mode."
_________

Articles are a type word called a determiner. Determiners are parts of noun phrases that provide information about specificity, quantity, or possession.  

"A" and "an" are indefinite articles. They are used before a singular countable noun when referring to any member of a group. For example, "Download a text editor." means any text editor will do.

> Note that "countable" nouns are just that: things that can be separated from a group, where the separated items can be counted.

Author: This is a case where, grammatically speaking, there should be a lowercase after a colon: when you are presenting a "definition" (the other is a list). However, the easiest thing to do here is to write around it. "Note that "countable" nouns are things that can be separated from a group and counted."

| You need a token (or *the* token, or *your* token) to access the channel. | You need token to access the channel. | 

Some types of nouns do not take articles. For example, non-specific plural nouns ("Script files can be useful."), proper nouns ("Agora is a wonderful company."), and uncountable or mass nouns ("Planning is crucial to success.").

Note that many nouns can shift between "countable" and "uncountable" based on context, and they may therefore sometimes require an article and sometimes not. For example, "Information can be lost to user error. Be sure to back up the information in your project folder."

#### Choosing "a" or "an"

The sound of a word or acronym when it is spoken determines whether "a" or "an" precedes it. Use "a" before a word that starts with a consonant *sound*, and "an" for words that begin with a vowel *sound*. For example "a URL" and "an SDK".

#### Articles and internal agreement

Methods and callbacks must be specified, so when referring to a method or callback, use a leading "the" followed by "method" or "callback". For example "the SDK triggers the Client.on("disable-local-video") callback".

> You can omit “the” and “method” or "callback" when the reader will understand what it is through context.

Author: Ugh. Curly quotes are a pain.

``` admonition:: Example

  The CheckPermission method is used to request access to the microphone. You do not need to call CheckPermission for earlier versions.
```

#### Use of "the" before a type of mode

In most cases, do not use a leading article for "modes." The exception to this is modes that have multiple options.

For example:

- silent mode
- dual-stream mode
- individual recording mode
- render mode
- video display mode

> The following are modes that have several options: //TODO William Is this still true? Lots of places on the Internet (Zoom, Microsoft) seem to just use "mirror mode," for example. I'm not sure what you mean by "several options". Why isn't in just "set your device to recording mode."?

Author: Hmm. I would personally prefer being specific without the article: "Set your device to individual recording mode." I think for situations where the option isn't chosen, that is when you need an article: "Set your device to the recording mode appropriate to your scenario." Does that make sense?

> The following are modes that have several options: //TODO William Is this still true? Lots of places on the Internet (Zoom, Microsoft) seem to just use "mirror mode," for example. I'm not sure what you mean by "several options". Why isn't in just "set your device to recording mode."?

Reviewer: Agreed.

- the mirror mode
- the recording mode
- the rendering mode

### Capitalization

**Quick reference:**
- Use sentence case for titles, headings, and captions. Exception: Also capitalize the first word after a colon.
- Use sentence case for all table elements, including column headings.
- Capitalize the first word after a colon in a sentence if it follows a label such as "Caution," "Note," or "Example."  
- Use the official capitalization of products, companies, and so on.
- Do not use all-caps for emphasis. Use italics instead.
- Use sentence case for list items.
- Capitalize only the first word in a hyphenated phrase. Example: "Random-access memory"
_________

> In document titles and headings, use sentence case, where you capitalize only the first word, plus any proper nouns (such as names, trademarks, keywords, and other terms that are normally capitalized). You should also capitalize the first word after a colon in titles and headings.

Author: Okay. If we are sticking with that, I'll fix it.

``` admonition:: Example

  Step one: Generate a builderToken
```

Use the official capitalization for the names of brands, companies, software, products, services, and so on. When in doubt, visit the offical web site of the company.

``` admonition:: Example

  Copy the Microsoft Word files to your iOS project folder.
```

> Do not use capitalization for emphasis, as this comes across as shouting to a native English reader. Use italics instead. //TODO William Correct?

Author: Okay.

"Which" is used to introduce non-restrictive clauses. A non-restrictive clause provides information that is useful, but not crucial to a sentence. For example, "Download the sample project, which you can find on GitHub." If you remove the clause "which you can find on GitHub," the sentence still conveys its main paint (the directive to "Download the sample project"). The non-restictive clause is useful information, but it is not crucial. Restrictive clauses are set off with commas.

### Contractions

**Quick reference:**
- Do not use contractions except in FAQ titles and headers.
_________

A contraction is a word or phrase that is shortened by dropping one or more letters. Examples include "let's" for "let us" and "can't" for "cannot." While native English readers understand them, they add unnecessary complexity for non-native readers. For example, contractions that end with the letter "s" can be mistaken for possessive nouns. In technical documentation, the use of contractions is discouraged because it sets an informal tone.

> The more conversational nature of an FAQ allows for the use of contractions, but restrict them to titles and headers. //TODO William Correct?

Reviewer: Yes

| Do this...          | ...instead of this      |
| :------------------- | :------------------- |
| Click on the Open button. A pop-up window opens. | Once you have clicked on the Open button, a pop-up window will open. |

#### Exceptions to using the present tense
The past tense is used for events that have happened definitively in the past and are unlikely to repeat. For Agora technical documenation, this is generally reserved for two situations: release dates and fixed issues.

Always frame a product's release date in the past tense, as the present tense is only correct on the day of release. Fixed issues described in release notes should also be written in the past tense to emphasize that the issues have been addressed.

| Do this...          | ...instead of this      |
| :------------------- | :------------------- |
| v3.0.0 was released on Mar 4, 2020. | v3.0.0 is released on Mar 4, 2020. |
| Disconnecting the microphone caused the audio to freeze. | Disconnecting the microphone causes the audio to freeze. |

> Avoid phrasing using *will* and *can*. //TODO William Why are you singling these out? "Will" is attached to the future tense, and "can" is conditional--perhaps there are occasions when it applies?

Author: Okay!

> Use the future tense only if there is a significant time delay that matters in context. For example, "Agora will grant you access to the service once your application has been approved." //TODO William I'm not sure about this example. Is there a better one?

Author: Agreed, but these are supposed to be the guidelines for when to actually use it. I will try to incorporate something closer to the Google rule: "Sometimes, of course, future tense is unavoidable because you're actually talking about the future (for example, The hedgehog will hibernate starting in October.)."

> The present prefect tense is used to emphasize a past event that has consequences that continue into the present. For example, "Get the information from the log file if error logging has been activated." The corresponding Chinese is 已经. //TODO William The Chinese translates to "already"--is this useful here?

Author: As with all verb tenses, the present perfect is appropriate when it is appropriate.  :-)  I think a general rule to use the present tense whenever possible, but to be allowed to use the past, present perfect, or even future tenses when necessary is the best guideline. A blanket rule to always use the present tense and nothing else is too constricting, because there will always be situations where that sounds awkward. I can try to write this section to make that more clear. Is that okay?

> The present prefect tense is used to emphasize a past event that has consequences that continue into the present. For example, "Get the information from the log file if error logging has been activated." The corresponding Chinese is 已经. //TODO William The Chinese translates to "already"--is this useful here?

Reviewer: Yes, please.

### Possessives

**Quick reference:**
- Do not use possessives with company or product names.
_________

For singular nouns, add "'s" to the end of the word to form a possessive.
For plural nouns that end in "s", like most English words, add an apostrophe on the end without the "s".
For plural nouns that do not end in "s", add "'s" to the end of the word. 

``` admonition:: Example

   Combine the teacher's video with each of the students' videos on the children's channel.
```

> Do not use a possessive with a product or company name. This is due to matters of trademark, because the "'s" may be mistaken for a part of the name, and because trademarks are often considered adjectives rather than nouns. //TODO William Maybe update? The Google Style Guide allows 's with a company name (but not with products). I am not sure why this distinction is made--perhaps they consider a company to be an entity that can "have things". I do not mind the Google rule, because something like "Agora's livestreaming products" or "Apple's latest iOS update" sound okay to me, and sometimes writing around the possessive to avoid it can sound clunky. However, I am okay with either rule, so you should just decide and let me know.

Reviewer: We have discussed this and decided to keep the current rule.

Author: Okay!

- Use the imperative mood for task steps.
- When writing references describing methods, use the third person (what it does) rather than the second person (what a developer would do with it).
- Use the first person voice for FAQ questions.

_________

Using the second person voice ("you") addresses your text to the reader, which is why it is also called direct address. Second person is more engaging to readers than third person. However, when using the second person you must make sure you are clear who your intended audience is.

Use the second person rather than including yourself by using the first person ("we"). Remember who your audience is and that you are in a different role. In Chinese documents, use "你" instead of "您". 

| Do this...                                          | ...instead of this                           |
| :--------------------------------------------------- | :---------------------------------------- |
| You can get an AppID using Agora Console.        | Developers can get an AppID using Agora Console.         |
| Once in GitHub, you can download the sample project. | Once in GitHub, we can download a sample project. |

> If you are describing something in the third person, avoid using "his," "her," or "he/she". You can rephrase the sentence, use the singular "they," or refer to the relevant role instead (in singular or plural form). //TODO William Is there a preference?

Reviewer: I think this depends on the context?

Author: Okay!

## Language and grammar

### Abbreviations, acronyms, and initialisms

**Quick reference:**
- Do not use abbreviations. Exceptions: Units of measure with numbers, and "Ltd." or "Inc."
- Capitalize acronyms.
- Do not use periods after the letters in acronyms. Exception: "U.S." when it is used as an adjective.

> - Spell out uncommon acronyms the first time they appear. Example: "ANA (A New Acronym)."

Reviewer: The example should be "A New Acronym (ANA)"

Reviewer: ```suggestion
- Spell out uncommon acronyms the first time they appear. Example: "A New Acronym (ANA)."
```

- Do not use Latin abbreviations such as "e.g." or "etc."
_________

Abbreviations are a shortened version of a word used to represent the whole. Examples include "s" for "seconds" or "approx." for "approximately." 

> As a general rule, only use abbreviations in technical documentation for units of measure, and then only when they are joined with a number. For example, "15 s" is acceptable, but "The interval is measured in s." is not. Do not use a period after an abbreviated unit of measure. For more information, see Units of Measure. 

Reviewer: ```suggestion
As a general rule, only use abbreviations in technical documentation for units of measure, and then only when they are joined with a number. For example, "15 s" is acceptable, but "The interval is measured in s." is not. Do not use a period after an abbreviated unit of measure. For more information, see [Units of Measurement](formatting-and-organization.md#units-of-measurement). 
```

In most other cases (such as "approx."), spell out the word. The only obvious exceptions are "Ltd." or "Inc." in a company's name.

Note that some short forms have entered common speech and are now accepted as words themselves, such as "app" or "demo." If you are unsure if a short form is in common use, you can check the dictionary or consult with other writers.

Acronyms and initialisms are a type of abbreviation formed from the initial letters of other words. The difference between them is in pronunciation; acronyms are pronounced like words, whereas with initialisms each letter is pronounced individually. Examples of acronyms include "GIF" for "Graphics Interchange Format" and "RAM" for "random-access memory." Examples of initialisms include "SDK" for "Software Development Kit" and "API" for "Application Programming Interface."

> The pronunciation is important when choosing which form of the indefinite article to use. For more information, see Articles. 

Reviewer: ```suggestion
The pronunciation is important when choosing which form of the indefinite article to use. For more information, see [Articles](language-and-grammar.md#articles-a-an-the). 
```

| Output by the Agora server   | Output by Agora's server  |

### Second person

**Quick reference:**
- Use the second person when writing technical documentation.
- If you need to use the third person, use gender-neutral language.
- Use the imperative mood for task steps.
- When writing references describing methods, use the third person (what it does) rather than the second person (what a developer would do with it).
- Use the first person for FAQ questions.

_________

Using the second person ("you") addresses your text to the reader, which is why it is also called direct address. Second person is more engaging to readers than third person. However, when using the second person you must make sure you are clear who your intended audience is.

Use the second person rather than including yourself by using the first person ("we"). Remember who your audience is and that you are in a different role. In Chinese documents, use "你" instead of "您". 

| Do this...                                          | ...instead of this                           |
| :--------------------------------------------------- | :---------------------------------------- |

> | You can get an AppID using Agora Console.        | Developers can get an AppID using Agora Console.         |

Reviewer: ```suggestion
| You can get an App ID using Agora Console.        | Developers can get an App ID using Agora Console.         |
```
